sdk_refactor_design_proposal.md

```diff
@@ -0,0 +1,392 @@
+                 login_id: str = None,
+                 ssl_verify: bool = True,
+                 url: str = None):
+
+  	# ...
+    
+def resources_list(self, list_constraints: dict = None) -> dict:
+    """
+    This method is used to fetch all available resources for the current
+    account. Results are returned as an array of identifiers.
+    """
+    params = {
+        'account': self._account
+    }
+    params.update(self._default_params)
+    if list_constraints is not None:
+        json_response = invoke_endpoint(HttpVerb.GET, ConjurEndpoint.RESOURCES,
+                                        params,
+                                        query=list_constraints,
+                                        api_token=self.api_token,
+                                        ssl_verify=self._ssl_verify).content
+    else:
+        json_response = invoke_endpoint(HttpVerb.GET, ConjurEndpoint.RESOURCES,
+                                        params,
+                                        api_token=self.api_token,
+                                        ssl_verify=self._ssl_verify).content
+
+    resources = json.loads(json_response.decode('utf-8'))
+
+    # Returns the result as a list of resource ids instead of the raw JSON only
+    # when the user does not provide `inspect` as one of their filters
+    if list_constraints is not None and 'inspect' not in list_constraints:
+        # For each element (resource) in the resources sequence, we extract the resource id
+        resource_list = map(lambda resource: resource['id'], resources)
+        return list(resource_list)
+```
+
+See the  [Client class](https://github.com/cyberark/conjur-api-python3/blob/main/conjur/api/client.py) and [API class](https://github.com/cyberark/conjur-api-python3/blob/main/conjur/api/api.py#L128) and for more details.
+
+1. Multiple responsibilities and not easily expandable.
+
+The *API* class handles multiple responsibilities. The API class handles the building of the request based on the types of information provided by the user *and* the invoking of the call *and* formatting the response.
+
+2. Areas of volatility are not encapsulated.
+
+We can add commands, change the way we present a certain command's output, and/or the types of parameters can manipulate the overall flow. Each command has its own associated function in `client.py` and `api.py` so each time we want to add a new command, we need to re-open multiple classes and add the desired functionality. This overcrowds the *Api*, allowing command-specific logic to spill into a very general class. 
+
+##### Proposed flow
+
+For these reasons, I propose we extract the command-specific logic into its own class. That way, the the *Api* class will be exposed to the specific details of a command, how to package the request, how to make the call, what format the request should be in. This works to isolate the areas of volatility should we require to support different use cases for a command. 
```

Comment:
Seems a bit over-engineering to me. you usually encapsulate logic that change together. To me it seems that all three stages (build, execute, format) will change together. I would still search for a library that can help doing the generic work of sending the request and formatting the response

Reply:
Can you provide an example of what you mean by encapsulating the logic that changes together?
I thought that this was what I was doing by encapsulating the 3 in their own resource class